File: dirkit.go
// # Dir Kit
//
// * A simple toolkit for folder and file handling that eliminates
// boilerplate or wraps commonly used functions in a consistent
// namespace for easy rememberance/importing.

package dirkit

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"time"
)

var safetyPath string = "D:/safety/" // Change on per-project needs

// Helper function for determining if a path exists on disk or not.
// Args:
//
//	path(string): The path to check
//
// Returns:
//
//	bool: True if the path exists on disk else false.
//	error: The fs.ErrNotExist error if the path does not exist else nil.
func pathExists(path string) (bool, error) {
	if _, err := os.Stat(path); errors.Is(err, fs.ErrNotExist) {
		return false, err
	}
	return true, nil
}

// A helper function to determine if a path is a directory not not.
// Args:
//
//	path(string): The path to check.
//
// Returns:
//
//	bool: True if the path is a directory else false.
//	error: A os.IsNotExist error if the path does not exists, else nil.
func isDir(path string) (bool, error) {
	fileInfo, err := os.Stat(path)
	if err != nil {
		if os.IsNotExist(err) {
			return false, err
		}
		return false, err
	}
	return fileInfo.IsDir(), nil
}

// Gets the content names, or full path for contents, of a directory.
// Args:
//
//	path(string): Directory path to list the contents of.
//	fullPath(bool): To return string names or full paths of directory contents.
//
// Returns:
//
//	[]string: String names or full paths of directory contents.
//	error: Any error created from attempting to read the directory, else nil.
func GetDirContents(path string, fullPath bool) ([]string, error) {
	var contents []string

	items, err := os.ReadDir(path)
	if err != nil {
		return make([]string, 0), err
	}
	for _, item := range items {
		var entry string
		if fullPath {
			entry = fmt.Sprintf("%s%s", path, item.Name())
		} else {
			entry = item.Name()
		}
		contents = append(contents, entry)
	}
	return contents, nil
}

// Creates a directory from teh given path.
// Args:
//
//	path(string): The directory path to create.
//
// Returns:
//
//	error: Any error created while attempting to create the directory, else nil.
func CreateDirectory(path string) error {
	exists, _ := pathExists(path)
	if !exists {
		err := os.Mkdir(path, 0777)
		if err != nil {
			return err
		}
	}
	return nil
}

// Creates a directory with today's date as the name.
// Args:
//
//	path(styring): The path to create the new folder in.
//
// Returns:
//
//	error: Any error created while attempting to create the directory, else nil.
func CreateDatedDirectory(path string) error {
	datePath := filepath.Join(path, GetDate())
	err := CreateDirectory(datePath)
	if err != nil {
		return err
	}
	return nil
}

// Deletes a directory and its contents as long as they are within the safety path.
// Args:
//
//	folderPath(string): The folder path to delete.
//
// Returns:
//
//	error: the *PathError created from os.RemoveAll if one was created, else nil.
func DeleteSafeDirectory(folderPath string) error {
	if strings.HasPrefix(folderPath, safetyPath) {
		err := os.RemoveAll(folderPath)
		if err != nil {
			return err
		}
		return nil
	}
	errorMsg := fmt.Sprintf("folder path is not within %s", safetyPath)
	return errors.New(errorMsg)
}

// Removes specified file as long as it is within the safety path.
// Args:
//
//	filepath(string): The path to the file you wish to delete.
//
// Returns:
//
//	error: A custom error if the filepath was not within the safety path or a *PathError err from
//	os.Remove, else Nil.
func DeleteSafeFile(filepath string) error {
	if strings.HasPrefix(filepath, safetyPath) {
		err := os.Remove(filepath)
		if err != nil {
			return err
		}
		return nil
	}
	errorMsg := fmt.Sprintf("file path is not within %s", safetyPath)
	return errors.New(errorMsg)
}

// Delete all files in a directory as long as they are within the safety path.
// Args:
//
//	directory_path(string): The path to the directory.
//
// Returns:
//
//	any *PathError crated from DeleteSafeFile or errors from GetDirContents, else nil.
func DeleteSafeFilesInDirectory(folderPath string) error {
	if strings.HasPrefix(folderPath, safetyPath) {
		files, err := GetDirContents(folderPath, true)
		if err != nil {
			return err
		}
		for _, file := range files {
			err := DeleteSafeFile(file)
			if err != nil {
				return err
			}
		}
		return nil
	}
	errorMsg := fmt.Sprintf("file path is not within %s", safetyPath)
	return errors.New(errorMsg)
}

// Copy file into a separate destination folder.
// Args:
//
//	source(string): File path of the file to copy.
//	dest(string): File path to copy the file too, optionally can have different name.
//
// Returns:
//
//	error: *PathError crated from os module or possible other error from io module else nil.
func CopyFile(source string, dest string) error {
	sourceFile, err := os.Open(source)
	if err != nil {
		return err
	}
	defer sourceFile.Close()

	destFile, err := os.Create(dest)
	if err != nil {
		return err
	}
	defer destFile.Close()

	_, err = io.Copy(destFile, sourceFile)
	if err != nil {
		return err
	}

	return nil
}

// Copy contents of a folder to the given destination.
// Args:
//
//	sourcePath(string): Folder path to the folder that is to be copied.
//	destination(string): Folder path to copy the folder + contents to.
//
// Returns:
//
//	error: Any relevant errors created durring process, usually os *PathErrors else nil.
func CopyFolderContents(sourcePath string, destination string) error {
	sourcePath = filepath.Clean(sourcePath)
	destination = filepath.Clean(destination)

	err := CreateDirectory(destination)
	if err != nil {
		return err
	}

	curItems, err := GetDirContents(sourcePath, false)
	if err != nil {
		return err
	}

	for _, item := range curItems {
		curItemPath := filepath.Clean(filepath.Join(sourcePath, item))
		destPath := filepath.Clean(filepath.Join(destination, item))

		dir, err := isDir(curItemPath)
		if err != nil {
			return err
		}
		if dir {
			err := CopyFolderContents(curItemPath, destPath)
			if err != nil {
				return err
			}
		} else {
			err := CopyFile(curItemPath, destPath)
			if err != nil {
				return err
			}
		}
	}
	return nil
}

// Returns string: 'yyyymmdd'.
func GetDate() string {
	return time.Now().Format("20060102")
}

// Returns string: 'HH:MM:SS:XX', X is microsecond.
func GetTime() string {
	return time.Now().Format("15:04:05:00")
}

// Exports a string map to json file path.
// Args:
//
//	fielpath(string): The file path to place the .json file.
//	data(map[string]interface{}): Any map with string keys and values that can be converted to strings.
//	overWrite(bool): To overwrite json file if it already exists in path.
//
// Returns:
//
//	error: Any relevant error from the json handling or file writing process.
func ExportMapToJson(filePath string, data map[string]interface{}, overWrite bool) error {
	exists, err := pathExists(filePath)
	if err != nil {
		return err
	}

	if !exists || overWrite {
		jsonData, err := json.Marshal(data)
		if err != nil {
			return err
		}

		file, err := os.Create(filePath)
		if err != nil {
			return err
		}
		defer file.Close()

		_, err = file.Write(jsonData)
		if err != nil {
			return err
		}

		return nil
	}
	return nil
}

File: dirkit.py
"""
# Dir Kit

* A simple toolkit for folder and file handling that eliminates
boilerplate or wraps commonly used functions in a consistent
namespace for easy rememberance/importing.
"""


import os
import re
import datetime
import shutil
import json
import platform
from typing import Optional
from typing import Union
from pathlib import Path


SAFETY_PATH = Path('D:/safety/')  # Change on per-project needs.


def get_dir_contents(path: Path, full_path: bool = False) -> Union[list[Path], list[str], None]:
    """
    Gets the content names, or full path for contents, of a directory.

    Args:
        path (pathlib.Path): Directory path to list contents of.

        full_path (bool): To return string names or paths of directory contents. Defaults to False.

    Returns:
        String names of directory contents if full_path = False,
        Paths for directory contents if full_path = True.
    """
    if path.exists():
        if full_path:
            contents = list(path.glob('*'))
        else:
            contents = os.listdir(path)

        return contents

    return None


def create_directory(path: Path) -> Path:
    """
    Creates a directory from the given path.

    Args:
        path (path): The directory path for the directory to create.

    Returns:
        str: The created, or pre-existing, directory path.
    """
    if not os.path.exists(path):
        os.makedirs(path)

    return path


def create_dated_directory(path: Path) -> Path:
    """
    Creates a directory with today's date as the name.

    Args:
        path (str): The path, with base directory name, to place the directory.

    Returns:
        str: The full path of the created directory with date.
    """
    date_path = Path(path, get_date())
    create_directory(date_path)

    return date_path


def delete_safe_directory(path: Path) -> None:
    """
    Deletes a directory and its contents as long as they are within the safety path.

    Args:
        path (Path): the path to the directory to delete. Will throw exception if
        path is not within the safety path.
    """
    if SAFETY_PATH in list(path.parents):
        for root, dirs, files in os.walk(path, topdown=False):
            for name in files:
                filename = os.path.join(root, name)
                os.remove(filename)
            for name in dirs:
                os.rmdir(os.path.join(root, name))

        os.rmdir(path)
    else:
        raise ValueError(f'Path must be within {SAFETY_PATH.as_posix()}!')


def delete_safe_file(filepath: Path) -> None:
    """
    Removes specified file as long as it is within the safety path.

    Args:
        filepath (Path): the path to the file to delete. Will throw exception if
        path does not start with the safety path.
    """
    if SAFETY_PATH in list(filepath.parents):
        os.remove(filepath)
    else:
        raise ValueError(f'Path must be within {SAFETY_PATH.as_posix()}!')


def delete_safe_files_in_directory(directory_path: Path) -> None:
    """
    Delete all files in a directory as long as they are within the safety path.

    Args:
        directory_path (Path): Path to the directory.
    """
    try:
        files = get_dir_contents(directory_path, True)
        for file in files:
            delete_safe_file(filepath=file)
        print("All files deleted successfully.")
    except OSError:
        print("Error occurred while deleting files.")


def copy_file(source: Path, destination: Path, new_name: Optional[str] = '') -> None:
    """
    Copy file into a separate destination folder.

    Args:
        source (Path): file path of the file to copy.

        destination (Path): folder path of where to copy the file to.

        new_name (Optional[str]): an optional argument to rename the file.
    """
    if source.parent == destination:
        if source.is_dir():
            return
        else:
            if '.' in new_name and not new_name.split('.')[-1].isnumeric():
                new_base_name = os.path.splitext(new_name)[0]
            else:
                new_base_name = new_name

            ext = source.name.split('.')[-1]
            replace_name = f'{new_base_name}.{ext}'
            shutil.copy(source, Path(source.parent, replace_name))
    else:
        create_directory(destination)

        if new_name:
            target = Path(destination, new_name)
        else:
            target = destination

        shutil.copy(source, target)


def copy_folder_contents(source: Path, destination: Path) -> None:
    """
    Copy contents of a folder to the given destination.

    Args:
        source (Path): folder path to the folder that is to be copied.

        destination (Path): folder path to copy the folder + contents to.
    """
    shutil.copytree(source, destination, dirs_exist_ok=True)


def get_date() -> str:
    """Returns str: 'YYYYMMDD'"""
    today = datetime.date.today()
    return today.strftime("%Y%m%d")


def get_time() -> str:
    """Returns str: 'HH:MM:SS:XX', X is microsecond."""
    now = datetime.datetime.now().time().isoformat()[:-4]
    return now


def get_os_info() -> tuple[str, str, str]:
    """Returns tuple[str, str, str]: OS name, release number, and version number."""
    system  = platform.system()
    release = platform.release()
    version = platform.version()
    return system, release, version


def export_data_to_json(path: Path, data: dict, overwrite: bool = False) -> None:
    """
    Export dict to json file path.

    Args:
        path (Path): the file path to place the .json file.

        data (dict|list): the data to export into the .json file.

        overwrite(bool): to overwrite json file if it already exists in path.
            Defaults to False.
    """
    if not path.exists() or overwrite:
        with open(path, 'w') as outfile:
            json.dump(data, outfile, indent=4)
    else:
        return


def import_data_from_json(filepath: Path) -> Optional[dict]:
    """
    Import data from a .json file.

    Args:
        filepath (Path): the filepath to the json file to extract data from.

    Returns:
        any: will return data if json file exists, None if it doesn't.
    """
    if os.path.exists(filepath):
        with open(filepath) as file:
            data = json.load(file)
            return data

    return None


def sort_path_list(path_objs: list[Path] = None) -> Optional[list[Path]]:
    """
    Alpha-numerically sorts a list of pathlib.Paths.

    Args:
        path_objs(list[Path]): The list of paths to sort.

    Returns:
        Optional[list[Path]]: The sorted list of paths or None if no
        list was provided.
    """
    if path_objs is None:
        return None

    if len(path_objs) == 1:
        return path_objs

    sort_strings = []
    for p in path_objs:
        sort_strings.append(p.as_posix())

    convert = lambda text: int(text) if text.isdigit() else text
    alphanum_key = lambda key: [convert(c) for c in re.split('([0-9]+)', key)]
    sort_strings.sort(key=alphanum_key)

    sorted_paths = []
    for s in sort_strings:
        sorted_paths.append(Path(s))

    return sorted_paths
